pkgs/development/haskell-modules/lib/compose.nix

```
@@ -730,4 +730,148 @@ rec {
+        «derivation /nix/store/r28dlymqid6q31dh5a46k36bag60ca52-witherable-0.5.drv»
+      ];
+      # ...
+      testHaskellDepends = [
+        «derivation /nix/store/rl0kfrl05z94in9f2fk8hl44r9pfyh44-HUnit-1.6.2.0.drv»
+        «derivation /nix/store/0dbqnww1fb3ml9f4s21phkw901yszd48-QuickCheck-2.14.3.drv»
+        # ...
+        «derivation /nix/store/0dml0xsw6s1v1d2qqhyk3qjyl4lw2fi4-vector-0.13.2.0.drv»
+      ];
+    }
+    ```
+  */
+  combineInputs =
+    {
+      # A list of packages to get the transitive inputs for.
+      #
+      # Note that while the `packages` parameter to `haskellPackages.shellFor`
+      # is a function from the Haskell package set to a list of packages, this
+      # function takes the list of packages directly.
+      packages,
+
+      # Extra dependencies, in the form of `haskellPackages.mkDerivation` build
+      # attributes.
+      #
+      # An example use case is when you have Haskell scripts that use
+      # libraries that don't occur in the dependencies of `packages`.
+      #
+      # Example:
+      #
+      #   extraDependencies = p: {
+      #     libraryHaskellDepends = [ p.releaser ];
+      #   };
+      extraDependencies ? { },
+
+      ...
+    }:
+    let
+      # This is a list of attribute sets, where each attribute set
+      # corresponds to the build inputs of one of the packages input to shellFor.
+      #
+      # Each attribute has keys like buildDepends, executableHaskellDepends,
+      # testPkgconfigDepends, etc.  The values for the keys of the attribute
+      # set are lists of dependencies.
+      #
+      # Example:
+      #   cabalDepsForSelected
+      #   => [
+      #        # This may be the attribute set corresponding to the `backend`
+      #        # package in the example above.
+      #        { buildDepends = [ gcc ... ];
+      #          libraryHaskellDepends = [ lens conduit ... ];
+      #          ...
+      #        }
+      #        # This may be the attribute set corresponding to the `common`
+      #        # package in the example above.
+      #        { testHaskellDepends = [ tasty hspec ... ];
+      #          libraryHaskellDepends = [ lens aeson ];
+      #          benchmarkHaskellDepends = [ criterion ... ];
+      #          ...
+      #        }
+      #        ...
+      #      ]
+      cabalDepsForSelected = builtins.map (p: p.getCabalDeps) (builtins.filter (p: p != null) packages);
+
+      # A predicate that takes a derivation as input, and tests whether it is
+      # the same as any of the `packages`.
+      #
+      # Returns true if the input derivation is not in the list of `packages`.
+      #
+      # isNotSelected :: Derivation -> Bool
+      #
+      # Example:
+      #
+      #   isNotSelected common [ frontend backend common ]
+      #   => false
+      #
+      #   isNotSelected lens [ frontend backend common ]
+      #   => true
+      isNotSelected = input: pkgs.lib.all (p: input.outPath or null != p.outPath) packages;
```

I think lib is in scope.
```suggestion
      isNotSelected = input: lib.all (p: input.outPath or null != p.outPath) packages;
```